src/hotspot/share/opto/mulnode.cpp

```diff
@@ -961,48 +962,72 @@ static bool const_shift_count(PhaseGVN* phase, Node* shiftNode, int* count) {
   return false;
 }
 
-static int maskShiftAmount(PhaseGVN* phase, Node* shiftNode, uint nBits) {
-  int count = 0;
-  if (const_shift_count(phase, shiftNode, &count)) {
-    int maskedShift = count & (nBits - 1);
-    if (maskedShift == 0) {
+// Returns whether the shift amount is constant. If so, sets real_shift and masked_shift.
+static bool mask_shift_amount(PhaseGVN* phase, const Node* shiftNode, uint nBits, int& real_shift, int& masked_shift) {
+  if (const_shift_count(phase, shiftNode, &real_shift)) {
+    masked_shift = real_shift & (nBits - 1);
+    return true;
+  }
+  return false;
+}
+
+// Convenience for when we don't care about the real amount
+static bool mask_shift_amount(PhaseGVN* phase, const Node* shiftNode, uint nBits, int& masked_shift) {
+  int real_shift;
+  return mask_shift_amount(phase, shiftNode, nBits, real_shift, masked_shift);
+}
+
+// Use this in ::Ideal only with shiftNode == this!
+// Returns the masked shift amount if constant or 0 if not constant.
+static int mask_and_replace_shift_amount(PhaseGVN* phase, Node* shiftNode, uint nBits) {
+  int real_shift;
+  int masked_shift;
+  if (mask_shift_amount(phase, shiftNode, nBits, real_shift, masked_shift)) {
+    if (masked_shift == 0) {
       // Let Identity() handle 0 shift count.
       return 0;
     }
 
-    if (count != maskedShift) {
-      shiftNode->set_req(2, phase->intcon(maskedShift)); // Replace shift count with masked value.
+    if (real_shift != masked_shift) {
       PhaseIterGVN* igvn = phase->is_IterGVN();
-      if (igvn) {
+      if (igvn != nullptr) {
         igvn->rehash_node_delayed(shiftNode);
       }
+      shiftNode->set_req(2, phase->intcon(masked_shift)); // Replace shift count with masked value.
     }
-    return maskedShift;
+    return masked_shift;
   }
+  // Not a shift by a constant.
   return 0;
 }
 
 // Called with
-//    outer_shift = (_ << con0)
+//   outer_shift = (_ << rhs0)
 // We are looking for the pattern:
-//   outer_shift = ((X << con1) << con0)
-//   we denote inner_shift the nested expression (X << con1)
-//
-// con0 and con1 are both in [0..nbits), as they are computed by maskShiftAmount.
+//   outer_shift = ((X << rhs1) << rhs0)
```

Comment:
Just an idea: To better keep track of what is the outer and inner rhs, we could use `rhs_inner` and `rhs_outer`.

Reply:
good idea!